### PR TITLE
Add a stop callback and indication if device is running

### PR DESCRIPTION
**Pull request by cameronmaske**
Hi @irmen 

This PR exposes miniaudio's `stopCallback` logic. Example below to show how it works!
```
def on_stop():
	print("Oh no, error!1")

capture = miniaudio.CaptureDevice()
...
capture.start(generator, on_stop)
```

I'm taken an opinionated approach here to only call the trigger the stop callback, if the device stops due to an error, not if stopped using the `stop` [method](https://github.com/irmen/pyminiaudio/blob/f453117f2469577d0ae268e332c77e62965dbb97/miniaudio.py#L1175). I guess technically, it's more of an error callback. 
Let me know if you have thought, happy to change this, or extend it (maybe pass a boolean to indicate if stop normally, or due to an error?).

I've also added a `.running` boolean to indicate if the capture/playback/duplex is has started, and has yet encountered a stop. This is a new useful feature I've been using in my own projects, to know if a stream dies. 
```
capture = miniaudio.CaptureDevice()
capture.start(...)
while capture.running:
   sleep(10)
```

Let me know if you have any feedback/thoughts. Happy to make changes as you see fit.

### CODE REVIEW

**Comment by irmen** [comment]
Thanks for making a PR!
What is the use of this callback?

Also, the PR as it is now, seems to break the unit test suite

**Comment by cameronmaske** [comment]
@irmen I have mainly used the callback to know if a device is disconnected (microphone/headphone are unplugged). 

Will try and figure out the test issues, looks like a `MA_NO_BACKEND` (`-103`) error is being raised.

File: .travis.yml [hunk 1]
```diff
@@ -13,9 +13,12 @@ matrix:
 dist: xenial
 sudo: false
 
+before_install:
+  - sudo apt-get update -qq
+  - sudo apt-get install -y jackd2 libjack-jackd2-0
```
Comment (irmen) [inline .travis.yml]: is this really  necessary

File: tests/test_miniaudio.py [hunk 2]
```diff
@@ -1,10 +1,58 @@
...
-        devs = miniaudio.Devices()
-        devs.get_playbacks()
-        devs.get_captures()
+
+def test_devices():
+    devs = miniaudio.Devices()
+    devs.get_playbacks()
+    devs.get_captures()
+
+def test_stop_callback_capture(backends, jackd_server):
```
Comment (irmen) [inline tests/test_miniaudio.py]: the tests currently still fail on python 3.5 it seems.

File: miniaudio.py [hunk 3]
```diff
@@ -1152,42 +1152,69 @@ def _internal_data_callback(device: ffi.CData, output: ffi.CData, input: ffi.CDa
...
+
+PlaybackCallbackGeneratorType = Generator[Union[bytes, array.array], int, None]
+CaptureCallbackGeneratorType = Generator[None, Union[bytes, array.array], None]
+DuplexCallbackGeneratorType = Generator[Union[bytes, array.array], Union[bytes, array.array], None]
+GeneratorTypes = Union[PlaybackCallbackGeneratorType, CaptureCallbackGeneratorType, DuplexCallbackGeneratorType]
+
 
 class AbstractDevice:
     def __init__(self):
+        self.running = False # type: Boolean
```
Comment (irmen) [inline miniaudio.py]: that type annotation should not be necessary?

File: miniaudio.py [hunk 4]
```diff
@@ -1152,42 +1152,69 @@ def _internal_data_callback(device: ffi.CData, output: ffi.CData, input: ffi.CDa
...
 
+@ffi.def_extern()
+def _internal_stop_callback(device: ffi.CData) -> None:
+    if not device.pUserData:
+        return
+    userdata_id = struct.unpack('q', ffi.unpack(ffi.cast("char *", device.pUserData), struct.calcsize('q')))[0]
+    callback_device = _callback_data[userdata_id]  # type: Union[PlaybackDevice, CaptureDevice, DuplexStream]
+    callback_device._stop_callback(device)
+
+PlaybackCallbackGeneratorType = Generator[Union[bytes, array.array], int, None]
```
[the remark this answers is not in the document]
Comment (cameronmaske) [reply miniaudio.py]: Good catch, this was the result of fast-forwarding an old commit (originally from the synthesizer repo) against the lastest master. Will revert.

**Comment by cameronmaske** [comment]
@irmen Thanks for the review. 

Just some context on the test changes to answer some of your comments on the review. 

Essentially, I've tried to add some integration tests to ensure everything is running as expected. 

It seems on TravisCI, the default Linux audio system ALSA, is very limited. There are no devices available, so starting something like capture or playback device, will just error out. After some unsuccessful attempts ([I tried this](https://github.com/travis-ci/travis-ci/issues/1754#issuecomment-497986438)) at creating a dummy device for ALSA, I came across this approach from another [python audio project](https://github.com/nocarryr/cython-sounddevice), that starts up a [jackaudio](http://jackaudio.org/) server with a dummy backend. 

So, I set up [pytest, to automatically start up a jackd server when needed for the stop tests](https://github.com/irmen/pyminiaudio/pull/10/files#diff-484462fced51d1a06b1d93b4a44dd535R24), and kill it once the tests run. For tests, [running on travis](https://github.com/irmen/pyminiaudio/pull/10/files#diff-484462fced51d1a06b1d93b4a44dd535R15) it will automatically default to using the Jack backend. If run locally, it will use the default backends. If the pytest fixture stuff is unfamiliar to you, I have [some short videos](https://www.cameronmaske.com/courses/introduction-to-pytest/watch/yield-pytest-fixture/), that explain what is going on. Essentially, they are just convenient ways to do setup/teardown functionality in tests. The `conftest.py` doesn't run any tests, it just houses these `fixtures` which can be shared with other tests. It's just a nice way separate re-usable setup/teardown logic from your `test_*.py` files and keeps those focused on actual tests. 

Regarding the test failing on python 3.5. It is failing to boot up jackd server, due to lack of locked memory. I'm not sure why this is different from the other environments. 
image
I've tried a bunch of different settings changes, but I can't seem to up the locked memory limit on travis
image
I'm not sure what's happening here. I don't think this is related to python 3.5 specifically, rather, the environment it runs on Travis appears to have a different memory configuration, which is causing the issue. 
Sorry about the excess of commits, but I tried debugging and reconfiguring it to have more memory, etc and with the lack of SSH access to the Travis instance, pushing commits is the only way to try out new settings. 

My workaround for now, [is to skip running the stop_callback py3.5 tests on Travis](https://github.com/irmen/pyminiaudio/pull/10/files#diff-29b41c89003764fb7619eb09f21a75eeR8). This is not ideal, but I'm not sure how to fix the issue atm.

File: miniaudio.py (hunk 3 above)
Comment (cameronmaske) [reply miniaudio.py]: I'm not too familiar with typing hints in python, so I may have this wrong. 
To my understanding, this helps other parts know to expect `self.running` should be a bool (previously I used the incorrect type of Boolean). 
Is this not the case?

Comment (irmen) [reply miniaudio.py]: The type analizer will see that is a boolean automatically because you assign False

**Comment by irmen** [comment]
Hi, there have been a lot of commits but the tests are in the green again. Is this now ready to be merged?  (we can squash everything so the number of commits doesn't matter)

File: miniaudio.py (hunk 3 above)
Comment (cameronmaske) [reply miniaudio.py]: Ah! Thanks for the clarification. Will remove the comment.

**Comment by cameronmaske** [comment]
@irmen Should be ready to be merged!

**Comment by irmen** [comment]
Thanks for the contribution

**Pull request merged**